### ISSUE TITLE
retain annotations at runtime

### ISSUE THREAD

**Issue by rmannibucau** [mapstruct/mapstruct#702]
@Mappings (and @Mapping) don't have a runtime retention. This is fine for mapstruct but I wanted to introspect the annotations in my project to generate a custom adoc report from the mapping class.

I see 2 solutions:

- mapstruct adds a report option/solution
- the retention is updated to let users rely on this model for their own purposes

**Comment by gunnarmorling** [comment]
Hi, couldn't you generate your report at build time, too, using an annotation processor? To me it seems like that'd be better suited for such once-off task (the mapper config cannot change at runtime anyways).

> mapstruct adds a report option/solution

What would such report entail?

**Comment by gunnarmorling** [comment]
Btw. if you feel adventurous, you could try and plug into the MapStruct processing chain. This will allow you to access the mapper model which is the basis for the code generation, so you could examine this to extract whatever information you need.

To do so, you'd have to implement [ModelElementProcessor](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/processor/ModelElementProcessor.java), register it via a `META-INF/services/...` file and add it to the processor path. It's not a public SPI at this point and I haven't tried it myself with external modell element processors, but it may be interesting to try.

**Comment by rmannibucau** [comment]
Report "constraint" is to not add another module and to not package the code. Thought to MEProc but didnt find a nice way to do it with mvn without multiple build phase but sounds like the best solution ATM.

**Comment by remmeier** [comment]
+1 it would be great if there is a standarized way of having all the evaluated mappings between source and target (not the raw annotations) available at runtime if desired.

we plan to use mapstruct to map JPA entities to DTO. As a consequence, we would like to do filtering, paging and sorting on DTOs. For this purpose, the paths/naming in DTO "queries" need to be translated back into entity "queries".

**Comment by remmeier** [comment]
I looked more closely at the processors. Would there be an interest for a PR that adds a  "MapperRuntimeModelRenderingProcessor" processor similar to the MapperRenderingProcessor that has its own set of templates to generate mapping informations (src <-> target properties) into the mapping classes? The processor could be enabled/triggered by an annotation like @RuntimeModel on the mapper class. Some further Model classes would be necessary at runtime to have an API to access the model. The thing could also be moved into a dedicated project.

**Comment by gunnarmorling** [comment]
Hi, generating some sort of metamodel for inspecting mappings is a very
interesting idea.

I like the idea of making this some kind of side project, so this may be
the right time to make the processor SPI more publicly available (currently
it's "internal"), probably marked as experimental/incubating.

That said, any ideas on how the generated metamodel would look like? A PR
would be welcome for sure. It'd be great if you could sketch a draft of the
generated code first, so we could discuss that before implementing it all
the way.

Thanks!

Am 22.07.2016 9:15 vorm. schrieb "Remo" <notifications@github.com>:

I looked more closely at the processors. Would there be an interest for a
PR that adds a "MapperRuntimeModelRenderingProcessor" processor similar to
the MapperRenderingProcessor that has its own set of templates to generate
mapping informations (src <-> target properties) into the mapping classes?
The processor could be enabled/triggered by an annotation like
@RuntimeModel on the mapper class. Some further Model classes would be
necessary at runtime to have an API to access the model. The thing could
also be moved into a dedicated project.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/mapstruct/mapstruct/issues/702#issuecomment-234471737>,
or mute the thread

**Comment by remmeier** [comment]
The processor model packages are quite extensive (which is a good thing for the processor impl). 
Not sure whether to expose part of it or work with a simplified runtime model that might be more accessible to consumers and easier to generate. In such a case, the processor model could be transformed into the runtime model. Then the code generation/serialization for the model part would be done with the runtime model. Or maybe the processor model could extend the runtime model.
But I'm a bit afraid of doing too many changes to the current processor impl to early. Maybe that could also be done in a second step to keep the implementation separate as add-on for the time being.  

Mapper classes could implement an interface to trigger the runtime model generation:

`org.mapstruct.model.MapperModelAware 
- getModel : org.mapstruct.model.MapperModel`

which would look like:

`@Mapper
public abstract class TestMapper implements MapperModelAware{
	public abstract TestDto testMap(TestEntity entity);
}`

First I tought of using Framemaker templates to do the model generation as well. But after looking how rich the model is, a serialization approach like http://x-stream.github.io/ might be better suited and is far easier to implement.

**Comment by sjaakd** [comment]
Do I understand you correctly that you initially would consider property mapping (as part of bean mapping) only? Would the way we implemented `BuiltIn` methods be a way forward or would that be too limited? Can that mechanism be published?

**Comment by remmeier** [comment]
yes, I guess the BeanMappingMethod would be the most important and initial one. In my case to map DTO sorting/filtering to JPA sorting/filtering. But other things like the conversion of enums would be fairly interesting as well very quickly. So I suspect the runtime model will grow a bit.

Similar to the BuiltIn methods sounds good.

**Comment by sjaakd** [comment]
One of the things I was thinking of to make the BuiltIn method api externally available and make it possible register your own methods (extending something extending 'BuiltIn' via the SPI.

Things that have to be considered are the order in which the matching mechanism works. The matching mechanism itself is perhaps also candidate for (SPI-like) extension. I.e. to add your own matching filters. Anyways.. food for thought.

Anyway.. Perhaps its to limited. WDYT: @gunnarmorling, @agudian ?

**Comment by julien-may** [comment]
What's the state here? Having a model that describes the mapping between a source and target class at runtime is indeed very useful when wanting/needing to separate concerns as described by @remmeier https://github.com/mapstruct/mapstruct/issues/702#issuecomment-232922486

**Comment by filiphr** [comment]
@julien-may No one has been working on this. In case anyone is interested to work on such feature we can discuss it more

**Comment by remmeier** [comment]
how about generating a simple properties file that specifies which pairs of target class/property is mapped from which source class/property? like

```
mypackage.MyTargetClass.myTargetProperty=mypackage.MySourceClass.mySourceProperty.nestedProperty
```

this would give those mapping to follow, for example, DTO mappings.

If I want to map an actual single value, I could setup an empty MySourceClass, fill in that one property and run it trough the regular mapper. Then I can extract it again on the target side. The combination of this and the properties would open up already many use cases. 

alternatively one could serialize BeanMappingMethod, etc. to XML to get a much more detailed model at runtime?

**Comment by rmannibucau** [comment]
Reflection is generally the easiest to integrate but parsing a file works even if way less handy at usage since it requires more work and conversion instead of just using an already loaded model. It is doable with asm too without files but not trivial as moving annotations retention.

**Comment by remmeier** [comment]
yes retention could be a first simple step. The one thing I don't like is that one has to repeat all the logic/work already done by MapStruct: decide whether there is a dedicated mapping, or just a default mapping, where there is a mapping from a shared configuration, etc. A simple service would be nice to have that is able to translate a property path to a mapped property path.

**Comment by rmannibucau** [comment]
Hmm, was not 100% the usage i was doing. I was indeed extracting mapping annotation config to document the mapper but as a passthrough (see mapstruct for details) and adding some @Documentation for business purpose with human oriented description. So mapping was more a marker. That said if we can have both a report generator and runtime retention it would be perfect :).

**Comment by TobseF** [comment]
We also had the problem to access the `Mapping` Annotations at runtime. 
We use [MapStruct](https://mapstruct.org/) between a _business-model_ and our _ui-model_. 
When a UI client wants to get sorted data, it can specify a field from the _ui-model_. 
Our `MapStructParser` can get the corresponding _business-model_ field-name and create the needed Criteria to sort it.

We solved this problem by using [ASM](https://asm.ow2.io/) (bytecode manipulation and analysis framework)
 to read the `Mapping`-Annotation and provide it in a meta model:

``` java
public class AnnotationParser {
   public void parse(Class<?> mapper) {
      ClassInfoCollector classPrinter = new ClassInfoCollector(annotationInfos);
      ClassReader cr = new ClassReader(mapper.getCanonicalName());
      cr.accept(classPrinter, 0);
   }
}

public class ClassInfoCollector extends ClassVisitor {

   private final List<MethodAnnotationInfo> mapStructParser;

   public ClassInfoCollector(List<MethodAnnotationInfo> mapStructParser) {
      super(ASMversion);
      this.mapStructParser = mapStructParser;
   }

   @Override
   public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
      super.visit(version, access, name, signature, superName, interfaces);
   }

   @Override
   public MethodVisitor visitMethod(int access, String methodName, String descriptor, String signature, String[] exceptions) {
      return new MethodInfoCollector(methodName,  mapStructParser);
   }

}

public class ClassInfoCollector extends ClassVisitor {

   private final List<MethodAnnotationInfo> mapStructParser;

   public ClassInfoCollector(List<MethodAnnotationInfo> mapStructParser) {
      super(ASMversion);
      this.mapStructParser = mapStructParser;
   }

   @Override
   public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
      super.visit(version, access, name, signature, superName, interfaces);
   }

   @Override
   public MethodVisitor visitMethod(int access, String methodName, String descriptor, String signature, String[] exceptions) {
      return new MethodInfoCollector(methodName,  mapStructParser);
   }

}

class MethodInfoCollector extends MethodVisitor {

	private final String methodName;
	private final List<MethodAnnotationInfo> mapStructParser;

   public MethodInfoCollector(String method, List<MethodAnnotationInfo> mapStructParser) {
      super(ASMversion);
      this.methodName = method;
      this.mapStructParser = mapStructParser;
   }

   @Override
   public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) { 
      return new MethodAnnotationInfoCollector(methodName, descriptor,  mapStructParser);
   }
}

class MethodAnnotationInfoCollector extends AnnotationVisitor {

   private final String method;
   private final String annotationType;
   private final List<MethodAnnotationInfo> mapStructParser;

   public MethodAnnotationInfoCollector(String method, String annotationType, List<MethodAnnotationInfo> mapStructParser) {
      super(ASMversion);
      this.method = method;
      this.annotationType = annotationType;
      this.mapStructParser = mapStructParser;
   }

   @Override
   public void visit(String name, Object value) {
      MethodAnnotationInfo annotationInfo = new MethodAnnotationInfo(method,  annotationType, name, value.toString());
      mapStructParser.add(annotationInfo);
      super.visit(name, value);
   }

}
```

Wit this `AnnotationParser` it's possible to get the mapping info in this way:

``` java
class BusinessModel{
   String zip; 
}

class UiModel{
   String plz; 
}

public interface ModelMapping extends BridgeMapping<BusinessModel, UiModel> {
   @Mapping(source = "zip", target = "plz")
   UiModel modelToUiModel(BusinessModel model, @MappingTarget UiModel uiModel);
}

@Test
public testMappingInfo(){
   MapStructParser mappingInfo = new MapStructParser();
   mappingInfo.parseMappingInterface(ModelMapping.class);
   assertEquals("zip", mappingInfo.mapToTargetField("plz"));
}
```
The `mappingInfo.mapToTargetField("plz")` returns the mapped field of if the `BusinessModel` (`zip`).  
The `AnnotationParser` is a general purpose annotation parser which provide a list of `MethodAnnotationInfo`.  
 The `MapStructParser` uses the `AnnotationParser` model to build a `MapStructMappingInfo` by collecting the `Mapping`-Annotations.

The full runnable and tested example is available in here:  
https://github.com/TobseF/mapstruct-metadata-example

Stackoverflow: https://stackoverflow.com/posts/62675890/edit

> Theoretically it also possible to hook into the MapStruct annotation processing process and generate MetaModel Java classes (`ModelElementProcessor`). But I wasn't able to get it working. Adding additional Annotation processors is't straight forward and debugging Annotation processing during class compilation is more than cumbersome. Four our purpose the ASM way and a simple mapping is sufficient.

**Comment by techexplorer0310** [comment]
Is copying Mappings and Mapping annotation locally in to our application with the same package structure and changing the retention policy to RUNTIME an option? It would start using our overridden version of annotations?
